### ISSUE TITLE
Filter by column with "null" value (using ALLOW FILTERING) return empty data set

### ISSUE THREAD

**Issue by juliayakovlev** [scylladb/scylla#4776]
*Installation details*
Scylla version (or git commit hash): master (commit 3af8431a40ade56ae8e51f716c0556216b82808e)

It's impossible to filter by empty value in the column. The request doesn't return the data even such record exists. 

![Screenshot from 2019-07-29 15-29-00](https://user-images.githubusercontent.com/34435448/62116534-0728ec00-b2c3-11e9-9152-7ad20445366d.png)

**Comment by tzach** [comment]
Testing with Apache Cassandra 3.11.4

```
cqlsh:mykeyspace> CREATE TABLE hh (  pet_chip_id  uuid,  time timestamp,  heart_rate text, PRIMARY KEY (pet_chip_id, time));
cqlsh:mykeyspace> INSERT INTO hh(pet_chip_id, time) VALUES (123e4567-e89b-12d3-a456-426655440b55, '2019-08-04 07:31:00');
cqlsh:mykeyspace> INSERT INTO hh(pet_chip_id, time,heart_rate) VALUES (123e4567-e89b-14d3-a456-426655440b55, '2014-08-04 07:31:00', 'xxx');
cqlsh:mykeyspace> select * from hh;

 pet_chip_id                          | time                            | heart_rate
--------------------------------------+---------------------------------+------------
 123e4567-e89b-14d3-a456-426655440b55 | 2014-08-04 07:31:00.000000+0000 |        xxx
 123e4567-e89b-12d3-a456-426655440b55 | 2019-08-04 07:31:00.000000+0000 |       null

(2 rows)
cqlsh:mykeyspace> select * from hh WHERE heart_rate=null ALLOW FILTERING;
InvalidRequest: Error from server: code=2200 [Invalid query] message="Unsupported null value for column heart_rate"
```
Looks like a different bug

**Comment by juliayakovlev** [comment]
In Cassandra filtering by null still not supported. They mean to do that in 4.x version
https://issues.apache.org/jira/browse/CASSANDRA-10715

**Comment by dekimir** [comment]
Since 4f262e31d29a8af90908e5ba610040ec4df052aa, Scylla also forbids `WHERE heart_rate=null`.  It will become allowed again after we re-engineer the filtering code.

**Comment by AxelGreen** [comment]
@dekimir When can we expect that change?

**Comment by dekimir** [comment]
> @dekimir When can we expect that change?

@AxelGreen hopefully within days.  Watch #5763.

**Comment by avikivity** [comment]
@dekimir #5763 will only help people running master.

**Comment by dekimir** [comment]
> @dekimir #5763 will only help people running master.

@avikivity That was [our intent](https://github.com/scylladb/scylla/pull/5763#discussion_r405557323), wasn't it?

**Comment by avikivity** [comment]
"within a few days" is incorrect, for people who don't run master (which is most of them).

**Comment by dekimir** [comment]
> "within a few days" is incorrect, for people who don't run master (which is most of them).

Ah, OK, I see what you're saying.  OTOH, how many people run Scylla version between 4f262e3 and b0698df?  If I'm reading the tags correctly, that's just Scylla 4.1.0 and 4.1.1 ATM.

As for the original question "When can we expect that change?", I guess the answer highly depends on which Scylla version you currently run and which future version you're willing to upgrade to.

**Comment by dekimir** [comment]
And just to be clear: after #5763 lands, `WHERE heart_rate=null` will be allowed but still return no results.  Even if there are rows where `heart_rate` is missing, they will not be returned.  The new code considers any comparison involving NULL to be false, even when both arguments are NULL.

**Comment by nyh** [comment]
> And just to be clear: after #5763 lands, `WHERE heart_rate=null` will be allowed but still return no results. Even if there are rows where `heart_rate` is missing, they will not be returned. The new code considers any comparison involving NULL to be false, even when both arguments are NULL.

Wow, this doesn't sound very useful, and the people who ask for this feature (see also Casandra's https://issues.apache.org/jira/browse/CASSANDRA-10715 and our own #8489) want it to be useful.

**Comment by nyh** [comment]
> > And just to be clear: after #5763 lands, `WHERE heart_rate=null` will be allowed but still return no results. Even if there are rows where `heart_rate` is missing, they will not be returned. The new code considers any comparison involving NULL to be false, even when both arguments are NULL.
> 
> Wow, this doesn't sound very useful, and the people who ask for this feature (see also Casandra's https://issues.apache.org/jira/browse/CASSANDRA-10715 and our own #8489) want it to be useful.

On the other hand, it should be noted that in SQL, "where x = null" and "where x is null" is a different thing, and the former returns nothing (see https://stackoverflow.com/questions/9581745/sql-is-null-and-null). So our current behavior has some justification - although it will still be surprising for a lot of users, and, frankly, useless (why would anyone want to do a query which never returns any results)?

**Comment by nyh** [comment]
I created a cql-pytest test, `test/cql-pytest/run test_null.py::test_filtering_null_indexed`, for this feature.
As currently written, this test expects the feature as requested here by @juliayakovlev here and by Cassandra users in https://issues.apache.org/jira/browse/CASSANDRA-10715 - i.e., that `where v = null` should return rows where column `v` is unset. However, as explained above this is not the case today on either Scylla (matches nothing) or Cassandra (returns an error) so this test currently fails on both.

**Comment by dekimir** [comment]
> > And just to be clear: after #5763 lands, `WHERE heart_rate=null` will be allowed but still return no results. Even if there are rows where `heart_rate` is missing, they will not be returned. The new code considers any comparison involving NULL to be false, even when both arguments are NULL.
> 
> Wow, this doesn't sound very useful, and the people who ask for this feature (see also Casandra's https://issues.apache.org/jira/browse/CASSANDRA-10715 and our own #8489) want it to be useful.

That question was answered [here](https://github.com/scylladb/scylla/issues/7872#issuecomment-754648571).  How is this bug different from #7872?

**Comment by nyh** [comment]
> That question was answered [here](https://github.com/scylladb/scylla/issues/7872#issuecomment-754648571). How is this bug different from #7872?

It's indeed the same issue - #7872 was a duplicate of this older issue and nobody noticed.

But I'm afraid that closing that issue because "@avikivity thinks SQL does this" was a mistake, for two reasons I covered above:
1. The behavior that returns nothing is _not useful_, and nobody wants it to behave this way: We already had people open this issue (#4776), #7872, #8489, and Cassandra have https://issues.apache.org/jira/browse/CASSANDRA-10715 - that all wanted a different behavior.
2. Because SQL have a useless meaning for "= null", they also added "IS NULL" that does the "good" thing. But we don't have "IS NULL" (and "IS NOT NULL" is only implemented for MV definition), so people can't do what they want.

So I can see two good options to choose from now:
1. Add support for "IS NULL" and "IS NOT NULL" as in SQL, and whenever someone complains that "=null" doesn't work, tell them to use "is null" instead.
2. Change "=null" to do something useful - what people want it to do - regardless of what it does in SQL. 

For both options we need to do something. We can't just close this issue like we closed #7872.

**Comment by dekimir** [comment]
> But I'm afraid that closing that issue because "@avikivity thinks SQL does this"

First, it's not a matter of opinion -- SQL spec does explicitly cover that case.  And the actual reasoning was like this:
- SQL is explicit about NULL comparisons, while CQL isn't
- Postgres is consistent in treating NULL this way, Cassandra [isn't](https://github.com/scylladb/scylla/pull/5763#discussion_r405150654)
- therefore, users who come from SQL will have definite expectations, those from Cassandra won't
- therefore, it's better to mimic Postgres than Cassandra

> The behavior that returns nothing is not useful, and nobody wants it to behave this way

Does that mean SQL is not useful??  I suspect they had a good reason why they explicitly covered NULL comparisons in their standard.

> We already had people open this issue (#4776), #7872, #8489, and Cassandra have https://issues.apache.org/jira/browse/CASSANDRA-10715 - that all wanted a different behavior.

True, but is it better to mimic Postgres or to differ from both Postgres and Cassandra?

**Comment by dekimir** [comment]
> 1. Add support for "IS NULL" and "IS NOT NULL" as in SQL, and whenever someone complains that "=null" doesn't work, tell them to use "is null" instead.

I find this to be a much better option than the other, for reasons I outlined above.

Also, I suspect that users who use prepared statements would likely prefer the current `<=> null` behaviour than the second alternative you cite.

**Comment by nyh** [comment]
> First, it's not a matter of opinion -- SQL spec does explicitly cover that case. And the actual reasoning was like this:
> 
>     * SQL is explicit about NULL comparisons, while CQL isn't
> 
>     * Postgres is consistent in treating NULL this way, Cassandra [isn't](https://github.com/scylladb/scylla/pull/5763#discussion_r405150654)
> 
>     * therefore, users who come from SQL will have definite expectations, those from Cassandra won't

It is a matter of opinion.. An opinion of what people expect to have CQL based on various other prior art outside of CQL.
Frankly, I don't understand who would **want** "WHERE v = NULL" to match nothing. Why would anyone ever deliberately run a query that is known to match nothing? As I noted, we already know of 4 issues where a user wanted "WHERE v = NULL" to return rows with null v. Nobody ever ever asked us "why doesn't where v = NULL return absolutely nothing like SQL"? ;-)

> Does that mean SQL is not useful?? I suspect they had a good reason why they explicitly covered NULL comparisons in their standard.

Even if we do want to get inspiration for this from SQL (and I'm not sure this is a good idea, but it seems @avikivity and you are leaning strongly in that direction), then we have to go all the way - if we make "= null" a useless restriction (matching nothing) like in SQL, we need to also copy from them the useful "IS null" that they added - precisely because "= null" is useless.  

> True, but is it better to mimic Postgres or to differ from both Postgres and Cassandra?

If we implement "where v = NULL" in the way that everybody wanted (unlike SQL), we would not differ from Cassandra - we will simply be implementing a feature that Cassandra hasn't yet implemented. And we would be implementing it in the same way that the Cassandra people plan to do it (see https://issues.apache.org/jira/browse/CASSANDRA-10715). 
So I think you actually got it backward: If we implement an SQL-like (match-nothing) "where v = NULL", in a year we can find ourselves with the patch in https://issues.apache.org/jira/browse/CASSANDRA-10715 being committed in Cassandra, and now Scylla really will really differ from Cassandra.

**Comment by nyh** [comment]
> > 1. Add support for "IS NULL" and "IS NOT NULL" as in SQL, and whenever someone complains that "=null" doesn't work, tell them to use "is null" instead.
> 
> I find this to be a much better option than the other, for reasons I outlined above.

Very well. The reason why I don't think that this option is the _better_ one is that, so far, it seems that every user who wanted to look for null columns complained why "= null" doesn't work - nobody as far as I can tell asked why doesn't "IS NULL" work. I bet that if that if we add "IS NULL", most users will still try "= NULL" first, and then need to ask or read documentation to figure out that they need to switch to "IS NULL". That violates the https://en.wikipedia.org/wiki/Principle_of_least_astonishment.

Nevertheless, I do agree that adding "IS NULL" is a valid solution - so if you and @avikivity prefer it, I will concede - and create an issue (replacing this one), documentation issue, and tests for that solution (and tests that say that the current behavior of "= NULL" matching nothing is the right one). 

> Also, I suspect that users who use prepared statements would likely prefer the current `<=> null` behaviour than the second alternative you cite.

I'm afraid I don't understand why. Why would a user ever want to run a prepared statement with a null parameter knowing that this particular query will do nothing - and do so very slowly? In such a case, the user can simply not run the prepared statement... Can you give me an example why for prepared statement, this weird behavior is better?

**Comment by dekimir** [comment]
> > Also, I suspect that users who use prepared statements would likely prefer the current `<=> null` behaviour than the second alternative you cite.
> 
> I'm afraid I don't understand why. Why would a user ever want to run a prepared statement with a null parameter knowing that this particular query will do nothing - and do so very slowly? In such a case, the user can simply not run the prepared statement... Can you give me an example why for prepared statement, this weird behavior is better?

Easy: when users use NULL to mean "doesn't exist, please never show this in the front end."  I'm sure there are other scenarios, too.

Also, your presumption that the user deliberately puts NULL on the RHS is unjustified.  Most front ends generate parameterized database queries.  The NULL may be there unintentionally, come from an outside source, or result from some rare execution path the user doesn't want to bother checking for.

And the case is stronger for non-equality comparisons: an empty set is probably the best result for `v < ?` that happens to be bound to NULL.  And if that's the standard for `<` and `>`, then treating `=` differently is inconsistent.

Finally, I don't see much evidence that _everybody_ wants v=NULL to return a non-empty result.  Some do, which is why they file bugs like this one.  But that doesn't mean everybody; in fact, it could be a minority of users.

**Comment by nyh** [comment]
> I'm sure there are other scenarios, too. Any application being ported from SQL will have this expectation baked in; no astonishment there.

Somehow, this never happened in https://issues.apache.org/jira/browse/CASSANDRA-10715. Nobody ever came to that issue saying "Hey, wait a minute, I want = NULL to return nothing - my SQL-ported application needs it!".
Cassandra is a good test for what people want, because they never implemented "= NULL" in any way, so you can see what people asked for without being confounded by people who just got used to what they have.

> And the case is stronger for non-equality comparisons: an empty set probably _is_ the best result for `v < ?` that happens to be bound to NULL. And if that's the standard for `<`, `<=`, `>`, and `>=`, then treating `=` differently is inconsistent.

I don't agree. I think that "v < NULL" should be an error, as the user can avoid sending such a match-nothing request. This is especially true in Scylla which (I think?) doesn't short-circuit this filters-that-match-nothing and result in slow filtering queries for no good reason.
Cassandra's proposed patch in https://issues.apache.org/jira/browse/CASSANDRA-10715 allows "= NULL" but continues to disallow "< NULL".

**Comment by dekimir** [comment]
> Somehow, this never happened in https://issues.apache.org/jira/browse/CASSANDRA-10715. Nobody ever came to that issue saying "Hey, wait a minute, I want = NULL to return nothing - my SQL-ported application needs it!".
> Cassandra is a good test for what people want, because they never implemented "= NULL" in any way, so you can see what people asked for without being confounded by people who just got used to what they have.

The number of people expressing any opinion in all the bugs you linked is a single-digit number.  Let's maybe not assume that's representative of all CQL users on the planet today, let alone tomorrow.

Lacking a new response from @avikivity, I'm ready to close this bug by implementing `IS NULL`.

**Comment by nyh** [comment]
@avikivity please weigh in this issue. Should we go with SQL's "`=null` matches nothing so we add `is null` to do the useful thing" or with what a few (very few) Scylla and Cassandra users asked for - that `=null` do the useful thing?
Once we have a decision, I can make sure the right issues are open, and fix my test to expect the decided behavior.

**Comment by avikivity** [comment]
IMO, we should match SQL. CQL is a dialect of SQL, and if we make it subtly different we break expectations. There are many, many more SQL users than CQL users so the future pool of users include many more SQL users than CQL users.

SQL's `WHERE x = NULL` may seem useless, but it brings consistent treatment of NULL (any operation with NULL yields NULL, except for explicit nullness test). If we change `WHERE x = NULL` to mean `x IS NULL`, what do we do for `WHERE x = y`? Does it match when both are NULL? It should not, according to the definition that NULL means "unknown" (two unknown values are not equal, it is unknown whether they are equal or not).

**Comment by nyh** [comment]
@avikivity @dekimir ok, I give up. I'll rewrite the test, and should close this issue and open a new one for adding "IS NULL" instead.

One last question:
If an unset value (null) is considered "unknown", what happens if you filter "where x != 3"? If I understood you correctly, rows where x is unset (null) will not be retrieved. Is this correct and expected? I guess I need to write another test for that...

**Comment by dekimir** [comment]
> If an unset value (null) is considered "unknown", what happens if you filter "where x != 3"?

Are you asking what the SQL standard says or what Postgres will return?  The standard says the result of this comparison is unknown.  And Postgres can be easily checked.  My understanding has been all along that we should mimic Postgres.

**Comment by nyh** [comment]
> > If an unset value (null) is considered "unknown", what happens if you filter "where x != 3"?
> 
> Are you asking what the SQL standard says or what Postgres will return? The standard says the result of this comparison is unknown. And Postgres can be easily checked. My understanding has been all along that we should mimic Postgres.

I was very surprised to learn (after @dekimir told me...) that neither Cassandra nor Scylla support the "!=" operator in filters... It's part of the syntax and allowed for example in LWT conditions, but not in filters. If we try to use it to filter, we get the error:
```
Unsupported "!=" relation: b != 0
```
So at least that additional strange case I asked about doesn't exit.

**Comment by nyh** [comment]
I'm closing this issue as "wontfix", and also removing the "cassandra 4.x compatibility" tag because the feature requested here was never committed into Cassandra 4.x (https://issues.apache.org/jira/browse/CASSANDRA-10715 is still open).

As explained above, the consensus decision was that it is a feature, not a bug, that `WHERE x = NULL` returns an empty data set. I opened a new issue, #8517, requesting to support a different syntax - `WHERE x IS NULL` to do what this issue originally tried to do with `= NULL`.

**Comment by avikivity** [comment]
> @avikivity @dekimir ok, I give up. I'll rewrite the test, and should close this issue and open a new one for adding "IS NULL" instead.
> 
> One last question:
> If an unset value (null) is considered "unknown", what happens if you filter "where x != 3"? If I understood you correctly, rows where x is unset (null) will not be retrieved. Is this correct and expected? I guess I need to write another test for that...

`NULL != 3` is NULL, and NULL isn't true, so NULL rows should not pass the filter.

**Comment by nyh** [comment]
I just realized that not only does "WHERE x = NULL" works but matches nothing, deliberately, so does "WHERE x > NULL". More strangely, while "WHERE x = NULL" works without specifying ALLOW FILTERING, the "WHERE x > NULL" does require ALLOW FILTERING. This makes little sense :-( I'll prepare tests...
